gstd/gstd_http.c

```diff
@@ -0,0 +1,423 @@
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <string.h>
+#include <gst/gst.h>
+#include <libsoup/soup.h>
+
+#include "gstd_ipc.h"
+#include "gstd_http.h"
+#include "gstd_parser.h"
+#include "gstd_element.h"
+#include "gstd_pipeline_bus.h"
+#include "gstd_event_handler.h"
+
+/* Gstd HTTP debugging category */
+GST_DEBUG_CATEGORY_STATIC (gstd_http_debug);
+#define GST_CAT_DEFAULT gstd_http_debug
+
+#define GSTD_DEBUG_DEFAULT_LEVEL GST_LEVEL_INFO
+
+
+struct _GstdHttp
+{
+  GstdIpc parent;
+  guint port;
+  gchar *address;
+};
+
+struct _GstdHttpClass
+{
+  GstdIpcClass parent_class;
+};
+
+G_DEFINE_TYPE (GstdHttp, gstd_http, GSTD_TYPE_IPC);
+
+/* VTable */
+
+static void gstd_http_dispose (GObject *);
+static GstdReturnCode gstd_http_start (GstdIpc * base, GstdSession * session);
+static GstdReturnCode gstd_http_stop (GstdIpc * base);
+static gboolean gstd_http_init_get_option_group (GstdIpc * base,
+    GOptionGroup ** group);
+static SoupStatus get_status_code (GstdReturnCode ret);
+static GstdReturnCode do_get (SoupServer * server, SoupMessage * msg,
+    char **output, GstdSession * session);
+static GstdReturnCode do_post (SoupServer * server, SoupMessage * msg,
+    char *name, char *description, char **output, GstdSession * session);
+static GstdReturnCode do_put (SoupServer * server, SoupMessage * msg,
+    char *name, char **output, GstdSession * session);
+static GstdReturnCode do_delete (SoupServer * server, SoupMessage * msg,
+    char *name, char **output, GstdSession * session);
+static void do_request (SoupServer * server, SoupMessage * msg,
+    GHashTable * query, GstdSession * session, verb request_verb);
+static void server_callback (SoupServer * server, SoupMessage * msg,
+    const char *path, GHashTable * query, SoupClientContext * context,
+    gpointer data);
+
+static void
+gstd_http_class_init (GstdHttpClass * klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+  GstdIpcClass *gstdipc_class = GSTD_IPC_CLASS (klass);
+  guint debug_color;
+  gstdipc_class->get_option_group =
+      GST_DEBUG_FUNCPTR (gstd_http_init_get_option_group);
+  gstdipc_class->start = GST_DEBUG_FUNCPTR (gstd_http_start);
+  object_class->dispose = gstd_http_dispose;
+  gstdipc_class->stop = GST_DEBUG_FUNCPTR (gstd_http_stop);
+
+  /* Initialize debug category with nice colors */
+  debug_color = GST_DEBUG_FG_BLACK | GST_DEBUG_BOLD | GST_DEBUG_BG_WHITE;
+  GST_DEBUG_CATEGORY_INIT (gstd_http_debug, "gstdhttp", debug_color,
+      "Gstd HTTP category");
+}
+
+static void
+gstd_http_init (GstdHttp * self)
+{
+  GST_INFO_OBJECT (self, "Initializing gstd Http");
+  self->port = GSTD_HTTP_DEFAULT_PORT;
+  self->address = g_strdup (GSTD_HTTP_DEFAULT_ADDRESS);
+}
+
+static void
+gstd_http_dispose (GObject * object)
+{
+  GstdHttp *self = GSTD_HTTP (object);
+
+  GST_INFO_OBJECT (object, "Deinitializing gstd HTTP");
+
+  if (self->address) {
+    g_free (self->address);
```

Comment:
Dispose may be called multiple times. You need to NULL out address as well. And I suggest using `finalize` instead of `dispose`. The former is only called once.